### PR TITLE
Report issue for each occurrence of duplication

### PR DESCRIPTION
**Pull request by ABaldwinHunter**
This change fixes a bug in which we reported only the first occurrence
(or sexp) of duplicated code in a set, and dropped the rest. The others
were included in other_locations but not reported as an independent issue.

@codeclimate/review 
cc @pbrisbin 

First step in breaking apart https://github.com/codeclimate/codeclimate-duplication/pull/78

### CODE REVIEW

File: lib/cc/engine/analyzers/violation.rb [hunk 1]
```diff
@@ -96,8 +85,8 @@ def fingerprint
         end
 
         def description
-          description = "#{check_name} found in #{occurrences} other location"
-          description += "s" if occurrences > 1
+          description = "#{check_name} found in #{(other_sexps.length)} other location"
+          description += "s" if other_sexps.length > 1
```
Comment (pbrisbin) [inline lib/cc/engine/analyzers/violation.rb]: Now that we had to bring back `#occurrences`, I think you can revert this too.

File: spec/cc/engine/analyzers/javascript/main_spec.rb [hunk 2]
```diff
@@ -91,8 +95,8 @@
           <a className='button button-primary full' href='#' onClick={this.onSubmit.bind(this)}>Login</a>
     EOJSX
 
-    result = run_engine(engine_conf).strip
-    issues = result.split("\0")
+    issues = run_engine(engine_conf).strip.split("\0")
```
Comment (pbrisbin) [inline spec/cc/engine/analyzers/javascript/main_spec.rb]: I'd back this out. It's not a *terrible* case of drive-by, but just to flex the "essential changes only" muscles.

File: spec/cc/engine/analyzers/ruby/main_spec.rb [hunk 3]
```diff
@@ -43,10 +45,43 @@ module CC::Engine::Analyzers
         expect(json["other_locations"]).to eq([
           {"path" => "foo.rb", "lines" => { "begin" => 9, "end" => 13} },
         ])
+
```
Comment (pbrisbin) [inline spec/cc/engine/analyzers/ruby/main_spec.rb]: Accidental whitespace.

File: spec/cc/engine/analyzers/ruby/main_spec.rb [hunk 4]
```diff
@@ -43,10 +45,43 @@ module CC::Engine::Analyzers
...
+              it 'returns true' do
+                expect(subject.js?).to be true
+              end
+            end
+        EORUBY
+
+        issues = run_engine(engine_conf).strip.split("\0")
+
+        expect(issues.length).to eq(3)
```
Comment (pbrisbin) [inline spec/cc/engine/analyzers/ruby/main_spec.rb]: I think we need to get some better coverage, and can probably do so with a spec on `Violations`. Here you assert we get three `Violation`s out, which is good -- but I'd like to also confirm some attributes on each violation, that the main and other locations are accurate mostly.

File: lib/cc/engine/analyzers/violation.rb (hunk 1 above)
Comment (ABaldwinHunter) [inline lib/cc/engine/analyzers/violation.rb]: :+1:

File: spec/cc/engine/analyzers/ruby/main_spec.rb (hunk 3 above)
Comment (ABaldwinHunter) [inline spec/cc/engine/analyzers/ruby/main_spec.rb]: :+1:

File: spec/cc/engine/analyzers/ruby/main_spec.rb (hunk 4 above)
Comment (ABaldwinHunter) [inline spec/cc/engine/analyzers/ruby/main_spec.rb]: Sounds good. Adding a spec

**Comment by ABaldwinHunter** [comment]
@pbrisbin Ready for re-review!

**Comment by ABaldwinHunter** [comment]
Removed any drive-by changes and added spec.

File: spec/cc/engine/analyzers/javascript/main_spec.rb [hunk 5]
```diff
@@ -93,6 +95,7 @@
 
     result = run_engine(engine_conf).strip
     issues = result.split("\0")
+
```
Comment (pbrisbin) [inline spec/cc/engine/analyzers/javascript/main_spec.rb]: Accidental change.

**Comment by pbrisbin** [comment]
This LGTM.

I fully expect this to have a (potentially strong) impact on grades. We're taking what was a single thing representing N instances of duplication with X remediation_points and now reporting N things each with X remediation_points. This means the total remediation_points for the duplication will go from X to X*N. I'm fine with this, we'll discuss this and tune points and grades separately via a new PR, but it should be explained in the ChangeLog.

**Comment by ABaldwinHunter** [comment]
:+1:  

The mass and point algorithm changes coming on the heels of this PR will also impact grades. The overall impact may be smaller grade impact on more files.
